### PR TITLE
Blocked cursor

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
Generally looks good, and I would be inclined to merge it into 9.6 once it was ready.
A few comments, but nothing significant.
* Please add a brief comment into the jira about the implementation - i.e. that it is the cursors (which are accessed single threaded) that have the blocked io, rather than the index - which does support random access.  Any issues with multiple cursors when smart stepping?  etc. etc.
* also needs converting from a draft PR, rebasing and PR title updating so that jira bot association works.

**Comment by ghalliday** [comment]
@jakesmith please take a look.  Also, would it be easy to enable this option in thor?

File: common/thorhelper/thorfile.cpp [hunk 1]
```diff
@@ -132,9 +132,7 @@ IKeyIndex *openKeyFile(IDistributedFilePart & keyFile, size32_t blockedIndexIOSi
                 Owned<IFileIO> iFileIO = iFile->open(IFOread);
                 if (nullptr == iFileIO)
                     throw makeStringExceptionV(0, "Failed to open index file %s", remotePath.str());
-                if (blockedIndexIOSize)
-                    iFileIO.setown(createBlockedIO(iFileIO.getClear(), blockedIndexIOSize));
-                return createKeyIndex(remotePath.str(), crc, *iFileIO, (unsigned) -1, false);
+                return createKeyIndex(remotePath.str(), crc, *iFileIO, (unsigned) -1, false, blockedIndexIOSize);
```
Comment (jakesmith) [inline common/thorhelper/thorfile.cpp:135]: I don't think anything currently calls thorfile.hpp:createKeyIndex with blockedIndexIOSize at the moment?
Mainly called from hthor - to handle in future PR?

File: fs/dafsserver/dafsserver.cpp [hunk 2]
```diff
@@ -2236,7 +2236,7 @@ class CRemoteIndexBaseActivity : public CRemoteDiskBaseActivity
         crc32.tally(sizeof(time_t), &modTimeTT);
         unsigned crc = crc32.get();
 
-        keyIndex.setown(createKeyIndex(fileName, crc, isTlk));
+        keyIndex.setown(createKeyIndex(fileName, crc, isTlk, 0));
```
Comment (jakesmith) [inline fs/dafsserver/dafsserver.cpp:2239]: future PR? we should pick up blockedIndex size here for plane involve here in dafilesrv context too.

File: roxie/ccd/ccdfile.cpp [hunk 3]
```diff
@@ -826,7 +826,7 @@ class IndexCacheWarmer : implements ICacheWarmer
         if (nodeType != NodeNone && !keyFailed && localFile && !keyIndex)
         {
             //Pass false for isTLK - it will be initialised from the index header
-            keyIndex.setown(createKeyIndex(filename, localFile->getCrc(), *localFile.get(), fileIdx, false));
+            keyIndex.setown(createKeyIndex(filename, localFile->getCrc(), *localFile.get(), fileIdx, false, 0));
```
Comment (jakesmith) [inline roxie/ccd/ccdfile.cpp:829]: this is within warmBlock - is it reading orderd blocks, should it be using a blocked read ahead too here?

File: roxie/ccd/ccdfile.cpp [hunk 4]
```diff
@@ -1059,6 +1060,9 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
             const char *kind = fdesc.queryKind();
             if (kind && streq(kind, "key"))
                 isKey = true;
+            StringBuffer planeName;
+            fdesc.getClusterLabel(pdesc->copyClusterNum(0), planeName);
+            blockedSize = getBlockedFileIOSize(planeName);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply roxie/ccd/ccdfile.cpp:1065]: we can add a new pre-calculated version for index block size.

[the remark this answers is not in the document]
Comment (jakesmith) [reply roxie/ccd/ccdfile.cpp:1065]: looks like this var. is not longer used

File: roxie/ccd/ccdfile.cpp [hunk 5]
```diff
@@ -3139,17 +3144,20 @@ class CResolvedFile : implements IResolvedFileCreator, implements ISafeSDSSubscr
                                 IPartDescriptor *remotePDesc = queryMatchingRemotePart(pdesc, remoteFDesc, partNo-1);
                                 part.setown(createPhysicalFile(subNames.item(idx), pdesc, remotePDesc, ROXIE_KEY, fdesc->numParts(), cached != NULL, channel));
                                 pdesc->getCrc(crc);
+                                StringBuffer planeName;
+                                fdesc->getClusterLabel(pdesc->copyClusterNum(0), planeName);
+                                blockedIOSize = getBlockedFileIOSize(planeName);
```
Comment (jakesmith) [inline roxie/ccd/ccdfile.cpp:3149]: we can introduce a precalculated version for getBlockedIndexIOSize ..

File: roxie/ccd/ccdfile.cpp [hunk 6]
```diff
@@ -3139,17 +3144,20 @@ class CResolvedFile : implements IResolvedFileCreator, implements ISafeSDSSubscr
...
                             }
                         }
                         if (part)
                         {
                             if (lazyOpen)
                             {
                                 // We pass the IDelayedFile interface to createKeyIndex, so that it does not open the file immediately
-                                keyset->addIndex(createKeyIndex(part->queryFilename(), crc, *QUERYINTERFACE(part.get(), IDelayedFile), part->getFileIdx(), false));
+                                keyset->addIndex(createKeyIndex(part->queryFilename(), crc, *QUERYINTERFACE(part.get(), IDelayedFile), part->getFileIdx(), false, blockedIOSize));
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply roxie/ccd/ccdfile.cpp:3157]: I see a 0 passed for blockSize to createKeyIndex in warmBlock, not sure how related.

**Review by jakesmith — COMMENTED** [review]
@richardkchapman @ghalliday - looks good in general, a few minor comments/questions.

>There also be a roxie configuration default.

what should the default be in Roxie. Should be worried about the overall impact on heap?

**Pull request closed without merging**